app/src/main/java/io/appium/uiautomator2/server/mjpeg/MjpegScreenshotStream.java

```diff
@@ -0,0 +1,165 @@
+import java.util.Iterator;
+import java.util.Locale;
+
+import io.appium.uiautomator2.model.internal.CustomUiDevice;
+import io.appium.uiautomator2.server.ServerConfig;
+import io.appium.uiautomator2.utils.Logger;
+import io.appium.uiautomator2.utils.ScreenshotHelper;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class MjpegScreenshotStream extends Thread {
+    private static final String BOUNDARY_STRING = 
+        "--BoundaryString\r\n" +
+        "Content-type: image/jpg\r\n" +
+        "Content-Length: %d\r\n\r\n";
+    private static final int NO_CLIENTS_CONNECTED_SLEEP_TIME_MS = 500;
+    private static final byte[] END = "\r\n\r\n".getBytes(UTF_8);
+    private static final UiAutomation UI_AUTOMATION =
+        CustomUiDevice.getInstance().getInstrumentation().getUiAutomation();
+    private final List<MjpegScreenshotClient> clients;
+    private boolean isStopped = false;
+
+    MjpegScreenshotStream(List<MjpegScreenshotClient> clients) {
+        this.clients = clients;
+    }
+
+    @Override
+    public void interrupt() {
+        this.isStopped = true;
+        super.interrupt();
+    }
+
+    @Override
+    public void run() {
+        while (!isStopped) {
+            boolean noClientsConnected;
+            synchronized (clients) {
+                noClientsConnected = clients.isEmpty();
+            }
+
+            if (noClientsConnected) {
+                SystemClock.sleep(NO_CLIENTS_CONNECTED_SLEEP_TIME_MS);
+                continue;
+            }
+
+            // how long each loop should take in milliseconds
+            long targetInterval =
+                Math.round((1.0f / ServerConfig.getMjpegServerFramerate()) * 1000.0f);
+            long start = System.currentTimeMillis();
+
+            // add some resilience to handle `getScreenshot` or any of its
+            // sub-routines raising an error. If an error occurs during
+            // [post-]processing the frame will be dropped
+            byte[] screenshotData;
+            try {
+                screenshotData = getScreenshot();
+            } catch (Exception e) {
+                Logger.error("Error getting screenshot: ", e);
+                screenshotData = new byte[0];
+            }
+
+            if (screenshotData.length > 0) {
+                synchronized (clients) {
+                    Iterator<MjpegScreenshotClient> clientsIterator = clients.iterator();
+                    while (clientsIterator.hasNext()) {
+                        MjpegScreenshotClient client = clientsIterator.next();
+                        if (client.isClosed()) {
+                            clientsIterator.remove();
+                            continue;
+                        }
+
+                        if (!client.isInitialized()) {
+                            client.initialize();
+                        }
+
+                        client.write(screenshotData);
+                    }
+                }
+            }
```

Comment:
here in else block it also makes sense to leave some logs

Reply:
I don't think the `else` block is entirely necessary here, because most of the failure modes will come from exceptions being raised / caught and there is already logging there. But just to cover our based I've added a message.